src/ZEO/ClientStorage.py

```diff
@@ -1465,11 +1474,12 @@ def _process_invalidations(self, tid, oids):
             if oid == self._load_oid:
                 self._load_status = 0
             self._cache.invalidate(oid, tid)
-        self._cache.setLastTid(tid)
 
         if self._db is not None:
             self._db.invalidate(tid, oids)
 
+        self._cache.setLastTid(tid)
+
```

Comment:
Yes, to address the same race that Julien pointed out in the context of setting _txn_time in https://github.com/zopefoundation/ZODB/pull/56

If a connection called lastTransaction() after the cache tid was set but before getting the invalidations, then it would proceed with the transaction without having properly invalidating the cache.

This PR mitigates this race in 2 ways:

1. It gets the storage lock in ``lastTransactuon``.

2. It sets the cache tid (and thus the lastTransaction result) after sending invalidations.  The race is unlikely, but if we lose, we lose safely by using a transaction start time that's too early.